Complete Pause/Resume Implementation

codeslaw opened the pull request:
Hello!
After looking around for pause/resume info on Velocity, and finding a few issues on the matter (https://github.com/julianshapiro/velocity/issues/14, https://github.com/julianshapiro/velocity/issues/694), I've built out what I believe to be a fairly robust implementation of the "pause" and "resume" commands. It follows the same pattern of application as the "stop" command, and works on both immediate and delayed calls, and includes functional tests.

The general implementation works as follows:
On pause:
    if delayed, clear delay timeout, store remaining timing information
    else,  flag call as paused, store completion percent

On resume: 
    if delayed, setTimeout using stored timing info
    else, flag call as resumed

On tick:
   if paused, skip
   if resumed, reset call timeStart to currentTime - (duration * percentComplete) and clear paused flag

It's not a small PR, but I did attempt to build it in the spirit of how the rest of the library had been structured. If you have any questions or suggestions, let me know! If you do want to merge, I can rebase into one commit if you prefer.

Rycochet commented on velocity.js:
```diff
@@ -2271,6 +2272,144 @@
+						}
+					});
+
+					/* Pause and Resume are call-wide (not on a per elemnt basis). Thus, calling pause or resume on a 
```
Comment typo :-)

codeslaw replied:
Fixed :)

Rycochet commented:
Got another related pull req in #314 - with global pause/resume.

Not got enough time to go over this completely properly, but looks like it'll do the job. I am sort of wondering if storing the percentComplete is the best way though - while storing it does make sense in some ways, it's also got an accuracy issue with converting it back into timings, would storing a "pausedAt" time be better (I've got a feeling that the accuracy would be pretty much irrelevant for short animations, but what about long lasting ones - at what point might it become noticeable)?

Related to this - what happens if an element is paused, and another animation (with `queue:false`) is added?

One idea that's been bouncing around for a little while is being able to change the speed and position of individual animations (like a scrubber for a video) - so if you can think how to extend this code with that then do say :-)

codeslaw commented:
Thanks for the quick response! 

I did look at #314, but I don't think a global pause/resume satisfies the more compelling uses cases for pause/resume. Also, that PR seems to have been abandoned?

Good question on the use of percentageComplete. I think we are safe to use it here for a couple of reasons. The margin of error we'd be dealing with when converting timestamps to percentages and back is  0 < e < .0000001 milliseconds, even with durations in the trillions of ms. So at worst, we'd end up with a 1 frame difference in total animation duration in a paused vs. an un-paused animation of the same duration and starting time (think about a post-resume calculated startTime of 12434.00000001). However, by rounding that newly calculated value, we eliminate the error for any calls that have integer durations. I added a commit to do just that.

I also added tests to demonstrate the queue:false behavior. TLDR: queue:false animations will run immediately, queued animations will wait until the paused animation completes.

Another benefit from using percentageComplete is that we can extend it to support "speed" and "position" commands pretty easily. Here is a **super** simplified example

On position:
- previousPercentComplete = {0 <= p <= 1}
On tick:
- percentComplete = {0 <= previousPC + (currentPC - previousPC) * speedMultiplier <= 1}
On speed:
- speed = {real number}

Rycochet commented:
The ping for 314 was because if adding one, need to add both together - it's also against the older codebase, so needs to be manually merged and fixed together :-)

I definitely like this. Again, precision is relevant - the rounding might be ok, but I don't like using `Math.round()` simply because it means that something might be inconsistent - `Math.floor()` is my preferred way as at least that's consistent. Not sure if the high precision timer from `requestAnimationFrame()` is relevant here, not least because it's optionally turned off now (before it was always set to `Date.now()`. Will need to think about this more, but feels like it's at 99.9% - just something small I'm missing (might be better if I get time to look at this in a full source view on here - sadly that's likely to be the weekend).

As far as the global pause / resume goes - think it needs that simply to allow better control of things - I can see a use case of setting up a series of animations in Promises - which could conceivably be on separate ticks due to setImmediate style calls (there's a limit allowed in various browsers) - and wanting to start them all at the same time - or just pause *everything* in one call, without wanting to manually cycle through yourself - might want a polyfill against this for that one for consistency again ;-)

Finally - want to push the current build to a new minor release before pulling this (there's the text based forcefeeding which has earnt it above a patch version) - so hopefully be able to check through and get this pulled and global added simultaneously at the weekend.

codeslaw commented:
All that sounds good. And I agree re: accuracy. I went with Math.round because the floating point error can be positive or negative, so we'll get values of both 1234.00000000001 and 1233.99999999999. But yes, take your time and give it a review, and let me know what you think!

If you want to include global pause I'm all for it, but 314 didn't take into creating new anims while paused, animations in mid-delay. This branch has all that logic and could be refactored pretty trivially to achieve that goal. 

If you're cool with that I can re-factor 314 into this branch. Let me know!

Rycochet commented:
That makes sense, but what if the error is on something that's already 0.5 etc? Maybe add several more digits to it before rounding? Multiplication is cheap, division isn't - and we're sticking with performance - https://jsperf.com/ is back, so don't forget to test ideas :-)

Refactor away - as that's pretty much what I'm intending now - would let you pause everything and selectively resume, or pause selectively and let you resume everything ;-)

Rycochet commented:
Just had a thought - rather than storing percent, how about by frame count? Given that rAF is (supposed to be) 60fps, it would be `Math.ceil((end - start) * 60)` and `Math.ceil((now - start) * 60)` - using ceil as it checks that the end time has passed before completing - that way the accuracy issue could be bypassed by being per-frame (integer) accurate instead of perfect decimal...

codeslaw commented:
> ...would let you pause everything and selectively resume, or pause selectively and let you resume everything ;-)

Agreed! I'll take a stab at it and push something up soon.

Re: using frame count - Interesting thought, though any variance in frame rate may muck up the frame count. An easy compromise may be to track **millisecondsElapsed** instead of frameCount.  That gives us a useful relative value, and we still get the accuracy benefits you mentioned 😄

Rycochet commented:
I'm not thinking about *tracking* the frame count - just saying "ok, at this point in time we should have done xxx frames" - per the spec it's recommended to be 60fps, so even if it's actually done less (or more) that is accurate enough - and as all the running animations will get the same numbers, then starting and stopping will give a consistent behaviour. Basically the same end effect but rounding to frame boundaries rather than decimal boundaries.

You realise this is the same sort of issue that caused the change to use the rAF high precision timer by default? :-P

codeslaw commented:
Hey. I did some refactoring to get the global controls integrated.

Namely, I stopped tracking percentageComplete, started tracking msElapsed (per call). This is most analogous to your initial suggestions of using pausedAt timestamps, with a little less arithmetic. Bottom line is that it should now behave exactly the same way as the original engine in determining the most accurate representation of it's state.

Also a little bit of refactoring for code re-used on global and per-call pause/res commands.

I added another test suite, which is currently failing. There's likely a bug in pausing delayed calls globally, and also need to add several more tests. Normally I wouldn't push partially complete code, but won't be online for most of this weekend (though I may jump on late tonight/tomorrow). I just wanted to give you a heads up on the state of the branch.

Have a good weekend!

Rycochet commented:
That's cool - I'll have a look through it, and likely pull (to a branch maybe) + fix that use-case if I get time this weekend (rarely have time during the week, and I'm away all next weekend lol)

Rycochet commented:
Manually merged to try and figure out what's going wrong - looks like the global test itself was causing problems - even in unrelated tests, so disabled that, and everything looks good - started to make a go on updating the tests to get it runnnig, but ran out of time ;-)

codeslaw commented:
Glad to see that you pulled the code in, however, the way you manually merged it removed me entirely from the git history for the project.  I am totally in agreement on compressing it down to 1 commit before merging onto master, however, I would really like that 1 commit to have come from me. I put 25 hours into writing this, and it's important to me to be assigned credit for all of that work. 

I am currently working on fixing the global test, during which I have also found a bug in the global pause functionality. When I have that resolved, perhaps I can merge the entirety of the pause functionality into a single commit, and you could roll back the recent update to master, and then we merge from this PR?

Thanks for understanding! :-D

Rycochet commented:
Sounds like a plan - if you grab the changes from this end first, the test folder has changed somewhat, let me know when you're ready and we'll coordinate the rollback and merge :-P

I'm not going to have time to do anything else on the QUnit2 update this week (I think), so just the `asserts` are relevant right now.

codeslaw commented:
Sweet, I'll let you know!

Rycochet commented:
Reverted as there's other fixes needed - but can squash and merge from here - so just make sure it's up to date (don't worry about the .min as I tend to regenerate that anyway - really ought to get it happening automatically though...)

codeslaw commented:
Alright, I resolved a couple of global pause / resume issues and cleaned up the tests. Rebased off of the latest master and squashed to one commit. I didn't update the min. Should be good to go!

Rycochet commented:
Woot! :-D
